### PR TITLE
ConvertCPUSharesToCgroupV2Value: improve

### PR DESCRIPTION
**Pull request by kolyshkin**
The old formula, while correctly converting the cgroup v1 range to cgroup v2 range, had a few issues:

1. When cgroup v1 value is out of range, it returned invalid cgroup v2 value, leading to vague errors down the line (see [1]).
2. Default value cgroup v1 shares of 1024 is converted to 39, which is much less than the cgroup v2 default weight of 100 (see [2], [3]).

The new conversion formula fixes both issues (see discussion in [2] for more details).

Amend the test case accordingly.

[1]: https://github.com/opencontainers/runc/issues/4755
[2]: https://github.com/kubernetes/kubernetes/issues/131216
[3]: https://github.com/opencontainers/runc/issues/4772

### CODE REVIEW

**Comment by iholder101** [comment]
Thank you very much @kolyshkin!
/lgtm

**Review by AkihiroSuda — APPROVED** [review]

File: utils_test.go [hunk 1]
```diff
@@ -536,9 +536,12 @@ func TestGetHugePageSizeImpl(t *testing.T) {
...
-		2:      1,
-		262144: 10000,
+		0:      0,     // Unset.
+		1:      1,     // Below minimum (out of range).
+		2:      1,     // Minimum.
+		1024:   100,   // Default.
+		262144: 10000, // Maximum.
+		262145: 10000, // Above maximum (out of range).
 	}
 	for i, expected := range cases {
```
Comment (thaJeztah) [inline utils_test.go:546]: Not directly related to this PR, but we should use a different name for the `i` var to reduce cognitive load; it's very easy to read `i` as index value for a slice;

```suggestion
	for input, expected := range cases {
```

File: utils.go [hunk 2]
```diff
@@ -413,16 +414,28 @@ func WriteCgroupProc(dir string, pid int) error {
...
 	}
-	return (1 + ((cpuShares-2)*9999)/262142)
+	// Cgroup v1 CPU shares has a range of [2^1...2^18], i.e. [2...262144],
+	// and the default value is 1024.
+	// Cgroup v2 CPU weight has a range of [10^0...10^4], i.e. [1...10000],
+	// and the default value is 100.
+	if cpuShares < 2 {
+		return 1
+	}
+	if cpuShares > 262144 {
```
Comment (thaJeztah) [inline utils.go:431]: Really nitty, but these could be `<=` and `>=` to avoid doing the calculation

File: utils.go [hunk 3]
```diff
@@ -413,16 +414,28 @@ func WriteCgroupProc(dir string, pid int) error {
...
 func ConvertCPUSharesToCgroupV2Value(cpuShares uint64) uint64 {
+	// The value of 0 means "unset".
 	if cpuShares == 0 {
 		return 0
 	}
-	return (1 + ((cpuShares-2)*9999)/262142)
+	// Cgroup v1 CPU shares has a range of [2^1...2^18], i.e. [2...262144],
+	// and the default value is 1024.
+	// Cgroup v2 CPU weight has a range of [10^0...10^4], i.e. [1...10000],
+	// and the default value is 100.
```
Comment (thaJeztah) [inline utils.go:427]: I guess (see my other comment) this could be part of the GoDoc?

**Review by thaJeztah — COMMENTED** [review]
overall LGTM, but left some comments / suggestions.

File: utils_test.go (hunk 1 above)
Comment (kolyshkin) [reply utils_test.go]: Done (as a separate patch).

File: utils.go (hunk 2 above)
Comment (kolyshkin) [reply utils.go]: Done.

Although this also results in formula not being used for boundary values, so we're no longer testing it giving min and max as inputs. In order to test that we now need to add sharesMin+1 and sharesMax-1 values to the test.

Added as well.

File: utils.go (hunk 3 above)
Comment (kolyshkin) [reply utils.go]: Done

**Comment by kolyshkin** [comment]
@thaJeztah addressed all your comments, PTAL

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!!
